include/iommu.h

```diff
@@ -17,6 +17,128 @@
+#define IOMMU_MMIO_EXCLUSION_BASE	(0x20 >> 3)
+#define IOMMU_MMIO_EXCLUSION_LIMIT	(0x28 >> 3)
+#define IOMMU_MMIO_EXTENDED_FEATURE	(0x30 >> 3)
+#define IOMMU_MMIO_COMMAND_BUF_HEAD	(0x2000 >> 3)
+#define IOMMU_MMIO_COMMAND_BUF_TAIL	(0x2008 >> 3)
+#define IOMMU_MMIO_EVENT_LOG_HEAD	(0x2010 >> 3)
+#define IOMMU_MMIO_EVENT_LOG_TAIL	(0x2018 >> 3)
+#define IOMMU_MMIO_STATUS_REGISTER	(0x2020 >> 3)
+
+#define IOMMU_CR_IommuEn		(1ULL << 0)
+#define IOMMU_CR_HtTunEn		(1ULL << 1)
+#define IOMMU_CR_EventLogEn		(1ULL << 2)
+#define IOMMU_CR_EventIntEn		(1ULL << 3)
+#define IOMMU_CR_ComWaitIntEn		(1ULL << 4)
+#define IOMMU_CR_CmdBufEn		(1ULL << 12)
+#define IOMMU_CR_PPRLogEn		(1ULL << 13)
+#define IOMMU_CR_PprIntEn		(1ULL << 14)
+#define IOMMU_CR_PPREn			(1ULL << 15)
+#define IOMMU_CR_GTEn			(1ULL << 16)
+#define IOMMU_CR_GAEn			(1ULL << 17)
+#define IOMMU_CR_SmiFEn			(1ULL << 22)
+#define IOMMU_CR_SmiFLogEn		(1ULL << 24)
+#define IOMMU_CR_GALogEn		(1ULL << 28)
+#define IOMMU_CR_GAIntEn		(1ULL << 29)
+#define IOMMU_CR_DualPprLogEn		(3ULL << 30)
+#define IOMMU_CR_DualEventLogEn		(3ULL << 32)
+#define IOMMU_CR_DevTblSegEn		(7ULL << 34)
+#define IOMMU_CR_PrivAbrtEn		(3ULL << 37)
+#define IOMMU_CR_PprAutoRspEn		(1ULL << 39)
+#define IOMMU_CR_MarcEn			(1ULL << 40)
+#define IOMMU_CR_BlkStopMrkEn		(1ULL << 41)
+#define IOMMU_CR_PprAutoRspAon		(1ULL << 42)
+
+#define IOMMU_CR_ENABLE_ALL_MASK	(IOMMU_CR_IommuEn | \
+					 IOMMU_CR_HtTunEn | \
+					 IOMMU_CR_EventLogEn | \
+					 IOMMU_CR_EventIntEn | \
+					 IOMMU_CR_ComWaitIntEn | \
+					 IOMMU_CR_CmdBufEn | \
+					 IOMMU_CR_PPRLogEn | \
+					 IOMMU_CR_PprIntEn | \
+					 IOMMU_CR_PPREn | \
+					 IOMMU_CR_GTEn | \
+					 IOMMU_CR_GAEn | \
+					 IOMMU_CR_SmiFEn | \
+					 IOMMU_CR_SmiFLogEn | \
+					 IOMMU_CR_GALogEn | \
+					 IOMMU_CR_GAIntEn | \
+					 IOMMU_CR_DualPprLogEn | \
+					 IOMMU_CR_DualEventLogEn | \
+					 IOMMU_CR_DevTblSegEn | \
+					 IOMMU_CR_PrivAbrtEn | \
+					 IOMMU_CR_PprAutoRspEn | \
+					 IOMMU_CR_MarcEn | \
+					 IOMMU_CR_BlkStopMrkEn | \
+					 IOMMU_CR_PprAutoRspAon)
+
+#define IOMMU_EF_IASup			(1ULL << 6)
+
+#define COMPLETION_WAIT			1
+#define INVALIDATE_DEVTAB_ENTRY		2
+#define INVALIDATE_IOMMU_PAGES		3
+#define INVALIDATE_IOTLB_PAGES		4
+#define INVALIDATE_INTERRUPT_TABLE	5
+#define PREFETCH_IOMMU_PAGES		6
+#define COMPLETE_PPR_REQUEST		7
+#define INVALIDATE_IOMMU_ALL		8
+
+#define INVALID_CAP(c) 			((c == 0) || (c == 0xFFFFFFFF) || (c == 0xFF))
+
+/*
+ * Following are used to disable initial SLB protection only on
+ * platforms where DEV is present (Family 16h and older).
+ */
+#define DEV_PCI_BUS		0x0
+#define DEV_PCI_DEVICE		0x18
+#define DEV_PCI_FUNCTION	0x3
+
+#define DEV_OP_OFFSET		4
+#define DEV_DATA_OFFSET		8
+
+#define DEV_CR			4
+
+#define DEV_CR_SL_DEV_EN_MASK	1<<5
+
+/* Family 17h and newer */
+#define MCH_PCI_BUS		0x0
+#define MCH_PCI_DEVICE		0x18
+#define MCH_PCI_FUNCTION	0x0
+
+#define MEMPROT_CR		0x384
+
+#define MEMPROT_EN		1<<0
```

Comment:
This needs brackets, otherwise `~MEMPROT_EN` becomes `(~1)<<0`, same for `DEV_CR_SL_DEV_EN_MASK` above

Reply:
Indeed, I copied the problem to the new MEMPROT_EN define. So DEV_CR_SL_DEV_EN_MASK was already had the issue so how was this working? The disabling of DEV on the APU systems should have failed.

Reply:
It was explicitly bracketed in the point it was used: https://github.com/TrenchBoot/landing-zone/pull/76/files#diff-06fe7dc01e61244778e4a79b20e826b83073d3dc18d726b5fa7a1ec5cf87a156R157

Doing it in the header file instead is safer, too many parentheses only hurt eyes but not the code.

Reply:
Ah yes I see. I guess I should make everything consistent.